### PR TITLE
SONARJS-649 False Positive NonEmptyCaseWithoutBreak: complex flow

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 1]
```diff
@@ -48,26 +60,97 @@
...
+  private void addCaseExpression(Set<Tree> caseExpressions, ExpressionTree expression) {
+    caseExpressions.add(expression);
+    if (expression.is(Kind.CONDITIONAL_OR)) {
+      BinaryExpressionTree binary = (BinaryExpressionTree) expression;
+      // The left operand of a "||" is a valid predecessor for switch clause statements
+      addCaseExpression(caseExpressions, binary.leftOperand());
+    }
+  }
+
+  private boolean hasNonEmptyStatement(SwitchClauseTree nonEmptyCaseClause) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: MAJOR Make "hasNonEmptyStatement" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 2]
```diff
@@ -48,26 +60,97 @@
...
+        previousClauseWithStatement = switchClause;
+      }
+
+    }
     super.visitSwitchStatement(tree);
   }
 
-  private boolean endsWithJump(List<StatementTree> statements) {
-    if (statements.isEmpty()) {
+  private void addCaseExpression(Set<Tree> caseExpressions, ExpressionTree expression) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: MAJOR Make "addCaseExpression" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 3]
```diff
@@ -48,26 +60,97 @@
...
+  private static void addCaseExpression(Set<Tree> caseExpressions, ExpressionTree expression) {
+    caseExpressions.add(expression);
+    if (expression.is(Kind.CONDITIONAL_OR)) {
+      BinaryExpressionTree binary = (BinaryExpressionTree) expression;
+      // The left operand of a "||" is a valid predecessor for switch clause statements
+      addCaseExpression(caseExpressions, binary.leftOperand());
+    }
+  }
+
+  private static boolean hasNonEmptyStatement(SwitchClauseTree nonEmptyCaseClause) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas parameter name is weird

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 4]
```diff
@@ -48,26 +60,97 @@
...
+      addCaseExpression(caseExpressions, caseClause.expression());
     }
 
+    ControlFlowGraph cfg = getControlFlowGraph(tree);
+    SwitchClauseTree previousClauseWithStatement = null;
+
+    for (SwitchClauseTree switchClause : tree.cases()) {
+
+      if (previousClauseWithStatement != null
+        && (canBeFallenInto(switchClause, cfg, caseExpressions) || !hasNonEmptyStatement(previousClauseWithStatement))) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas this double negation is confusing. Can we remove it?

Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena Changed in 1978548.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java (hunk 3 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena Changed in 1978548.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 5]
```diff
@@ -48,26 +60,97 @@
...
+  }
+
+  private static boolean hasNonEmptyStatement(SwitchClauseTree nonEmptyCaseClause) {
+    return firstStatement(nonEmptyCaseClause.statements()) != null;
+  }
+
+  private static boolean canBeFallenInto(SwitchClauseTree switchClause, ControlFlowGraph cfg, Set<Tree> caseExpressions) {
+    StatementTree firstStatement = firstStatement(switchClause.statements());
+    if (firstStatement == null) {
+      // There's no block in the CFG for the statements of this switch clause
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas Confusing explanation: which statements are talking about (it looks like there are no statements)?

Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena We have to cover cases like `case x: {} {} ;;;`: there are statements here, but only "empty", and the CFG has no block for that. Any suggestion? Should we just drop the comment?

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas I suggest to rename method `firstStatement` into e.g. `firstNonEmptyStatement` and drop comment

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java [hunk 6]
```diff
@@ -48,26 +60,97 @@
...
 
+    ControlFlowGraph cfg = getControlFlowGraph(tree);
+    SwitchClauseTree previousClauseWithStatement = null;
+
+    for (SwitchClauseTree switchClause : tree.cases()) {
+
+      if (previousClauseWithStatement != null
+        && (canBeFallenInto(switchClause, cfg, caseExpressions) || !hasNonEmptyStatement(previousClauseWithStatement))) {
+
+        addIssue(previousClauseWithStatement, MESSAGE);
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @pynicolas Let's raise an issue on SwitchClauseTree#keyword() only. It doesn't look good now in UI

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java (hunk 5 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena Changed in 7f2791a

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java (hunk 6 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonEmptyCaseWithoutBreakCheck.java]: @vilchik-elena Changed in 73f5001

**Pull request merged**